### DISCUSSION TITLE
How would i mark malware PDF Files with clickable objects masquerading as filehoster?

### DISCUSSION THREAD

**Discussion by yoshimo** [rspamd/rspamd#4527] [Q&A]
There is a high amount of legitimate PDF files on this mail server with orders and receits and i can't plain say "block this kind of attachment".
Qakbot and other malware sends pdfs that tell you that there is a password protected download waiting for you and often use images from Office365, Google or other known Filehosters to be more legitimate.
Now i was wondering if we can tag e-mails whose pdf attachments have clickable links 

PDFID by Didier Stevens has the following on a sample

```
obj 13 0
 Type: /Action
 Referencing: 

  <<
    /Type /Action
    /S /URI
    /URI (https://<deleted>/dt/dt.php)
  >>
```

**Comment by fatalbanana** [comment]
Rspamd extracts URLs from PDFs (only of the clickable persuasion so far as I see) - and makes them available to be checked by the likes of `rbl` module. You could use a Lua rule which takes advantage of this:

```
rspamd_config.HAS_PDF_URL = {
  callback = function(task)
    local parts = task:get_parts()
    for _, part in ipairs(parts) do
      local ext = part:get_detected_ext()
      if ext == 'pdf' then
        local urls = part:get_urls()
        if urls[1] then
          return true
        end
      end
    end
  end,
  score = 1.0,
}
```
